crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1732 @@
+ */
+int ossl_ml_kem_encap_seed(uint8_t *ctext, size_t clen,
+                           uint8_t *shared_secret, size_t slen,
+                           const uint8_t *entropy, size_t elen,
+                           const ML_KEM_KEY *key)
+{
+    vinfo_t vinfo;
+    EVP_MD_CTX *mdctx;
+    int ret = 0;
+
+    if (!ossl_ml_kem_have_pubkey(key))
+        return 0;
+    vinfo = key->vinfo;
+
+    if (ctext == NULL || clen != vinfo->ctext_bytes
+        || shared_secret == NULL || slen != ML_KEM_SHARED_SECRET_BYTES
+        || entropy == NULL || elen != ML_KEM_RANDOM_BYTES
+        || key == NULL || (mdctx = EVP_MD_CTX_new()) == NULL)
+        return 0;
+
+    /*-
+     * This avoids the need to handle allocation failures for two (max 2KB
+     * each) vectors, that are never retained on return from this function.
+     * We stack-allocate these.
+     */
+#   define case_encap_seed(bits) \
+    case ML_KEM_##bits: \
+        { \
+            scalar tmp[2 * ML_KEM_##bits##_RANK]; \
+                                                                         \
+            ret = encap(ctext, shared_secret, entropy, tmp, mdctx, key); \
+            break; \
+        }
+    switch (vinfo->variant) {
+    case_encap_seed(512);
+    case_encap_seed(768);
+    case_encap_seed(1024);
+    }
+#   undef case_encap_seed
+
+    EVP_MD_CTX_free(mdctx);
+    return ret;
+}
+
+int ossl_ml_kem_encap_rand(uint8_t *ctext, size_t clen,
+                           uint8_t *shared_secret, size_t slen,
+                           const ML_KEM_KEY *key)
+{
+    uint8_t r[ML_KEM_RANDOM_BYTES];
+
+    if (key == NULL)
+        return 0;
+
+    if (RAND_bytes_ex(key->libctx, r, ML_KEM_RANDOM_BYTES,
+                      key->vinfo->secbits) < 1)
+        return 0;
+
+    return ossl_ml_kem_encap_seed(ctext, clen, shared_secret, slen,
+                                  r, sizeof(r), key);
+}
+
+int ossl_ml_kem_decap(uint8_t *shared_secret, size_t slen,
+                      const uint8_t *ctext, size_t clen,
+                      const ML_KEM_KEY *key)
+{
+    vinfo_t vinfo;
+    EVP_MD_CTX *mdctx;
+    int ret = 0;
+
+    /* Need a private key here */
+    if (!ossl_ml_kem_have_prvkey(key))
+        return 0;
+    vinfo = key->vinfo;
+
+    if (shared_secret == NULL || slen != ML_KEM_SHARED_SECRET_BYTES
+        || ctext == NULL || clen != vinfo->ctext_bytes
+        || (mdctx = EVP_MD_CTX_new()) == NULL) {
+        RAND_bytes_ex(key->libctx, shared_secret,
+                      ML_KEM_SHARED_SECRET_BYTES, vinfo->secbits);
+        return 0;
+    }
+
+    /*-
+     * This avoids the need to handle allocation failures for two (max 2KB
+     * each) vectors and an encoded ciphertext (max 1568 bytes), that are never
+     * retained on return from this function.
+     * We stack-allocate these.
+     */
+#   define case_decap(bits) \
+    case ML_KEM_##bits: \
+        { \
+            uint8_t cbuf[ML_KEM_##bits##_CIPHERTEXT_BYTES]; \
```

Comment:
Could almost just use MAX values here for the array sizes.

Reply:
Yeah, perhaps, but just in case stack space is precious when choosing to use ML-KEM-512, this is cheap enough to right-size.